pandas/core/internals/blocks.py

```diff
@@ -811,17 +821,24 @@ def _replace_coerce(
             if value is None:
                 # gh-45601, gh-45836, gh-46634
                 if mask.any():
-                    nb = self.astype(np.dtype(object), copy=False)
-                    if nb is self and not inplace:
+                    has_ref = self.refs.has_reference()
```

Comment:
IIUC checking self.refs.has_reference here vs a few lines down can have different results in the case where self.dtype == object?  does that make a difference?

if it doesnt make a difference, i think we can refactor this to share some code with some other CoW-handling